### ISSUE TITLE
Interval.after creates invalid interval when given a DateTime with microseconds and zero duration

### ISSUE THREAD

**Issue by Woodz** [moment/luxon#909]
**Describe the bug**
Given a DateTime that has a fractional millisecond component, `Interval.after` will create an invalid interval when given zero duration

**To Reproduce**
```
Interval.after(DateTime.fromMillis(123.456), 0)
```

**Actual vs Expected behavior**
Actual: `"The end of an interval must be after its start, but you had start=1970-01-01T07:30:00.123+07:30 and end=1970-01-01T07:30:00.123+07:30"`
Expected: Valid interval

**Desktop (please complete the following information):**

- OS: iOS
- Browser: Node 14.14.6
- Luxon version: 1.26.2
- TS: 4.2

**Additional context**
I am not sure if DateTime should be allowed to be constructed with a fractional millisecond component as it breaks a lot of other things (e.g. rendering as ISO without sub-milliseconds in the error message above)

**Comment by diesieben07** [comment]
I agree, fractional milliseconds should definitely be disallowed. Luxon uses many the browser Date APIs behind the scenes and they cannot deal with them - and as such Luxon cannot either.
An example:
```js
new Date(123.456).getUTCMilliseconds(); // 123
```
If you pass fractional milliseconds in, Luxon will keep those and `valueOf` will continue returning them. But a DateTime object also contains deconstructed date/time components (hours, seconds, etc.) for operations such as `plus`. These will have lost the fractional milliseconds. So by creating a DateTime with fractional milliseconds you have created a DateTime object which has inconsistent internal state. This should be disallowed.

There are two ways to go about this:

1. Use Math.trunc like `new Date` does
2. Create an invalid DateTime

I'm learning towards option 2.

**Comment by esetnik** [comment]
I'm not sure if this is related but I am also seeing some errors when `Interval.after` ends up calculating an end range in a different timezone and generates an error as a result. For example:

Invalid Interval: end before start: The end of an interval must be after its start, but you had start=2024-04-10T12:00:00.000-06:00 and end=2024-04-10T12:30:00.000-05:00

```js
  const dateLocal = DateTime.now();
  const duration = 1800000;
  const scheduledInterval = Interval.after(
    dateLocal,
    item.duration
  );
```

**Comment by diesieben07** [comment]
@esetnik Please clarify which time zone you are in. Judging by the ISO timestamps, it has to be something with UTC-6 in April and DST, which sounds like "Mountain Time", i.e. for example America/Denver.
Using this, I cannot reproduce your issue:
```js
const start = DateTime.fromISO('2024-04-10T12:00:00.000', {zone: 'America/Denver', setZone: true});
const duration = 1800000;
const scheduledInterval = Interval.after(
  start,
  duration
);
```

This code runs fine.

**Comment by esetnik** [comment]
This occurred in production to one of our users and i've been unable to reproduce it. However, I think the timezones shouldn't actually matter in this case and the function is doing the wrong thing. `Interval.after` provided with a positive duration should always produce a valid interval regardless of timezone changes / DST. Based on the offsets found in the error log I believe something related to CST / CDT is at play. However, according to wikipedia CST/CDT transition occurs in March and not April so the dates when this occurred don't correspond to a DST-related time change.

**Comment by diesieben07** [comment]
> `Interval.after` provided with a positive duration should always produce a valid interval regardless of timezone changes / DST.

Correct. And that is how it behaves, from how I understand it. I cannot reproduce the problem either, as I showed above.

**Comment by esetnik** [comment]
Ok, could a browser bug have caused this issue? I don't have any other information other than what's provided in our Sentry report. But one of our users definitely crashed on this and we have a check for positive duration on the line before so I don't think it's possible we provided a negative duration.

**Comment by diesieben07** [comment]
Do you have access to the client's time zone, the local time at the time of the issue and the browser? Those would help.

**Comment by esetnik** [comment]
Here's another example with as much detail as I can gather:

Crash occurred at: Jun 5, 2024 2:45:16 AM UTC according to the client browser
Error message: Invalid Interval: end before start: The end of an interval must be after its start, but you had start=2024-05-01T14:00:00.000-05:00 and end=2024-05-01T14:30:00.000-04:00
Error: Invalid Interval: end before start: The end of an interval must be after its start, but you had start=2024-05-01T12:00:00.000-05:00 and end=2024-05-01T12:30:00.000-04:00
Browser: Safari 17.4.1
System: Mac OS X >=10.15.7

**Comment by diesieben07** [comment]
Sorry, but without seeing more of the code that causes this issue, it is difficult to see where the issue lies.

**Comment by esetnik** [comment]
Hi I just wanted to report we are still seeing this issue in production. 26 instances across three users. I have no idea what could be going on.

**Comment by diesieben07** [comment]
If possible, please share as many details as you can about those 26 instances as well as the 3 users, so we can try to find something that they have in common.
Also, the more code you can share the better, as close to the actual code that runs as possible.

**Comment by esetnik** [comment]
FYI I am also able to reproduce this with the following code.

```ts
  // instead we move both start and end date to the local zone to ensure the
  // resulting interval uses the correct timezones
  const startDate = item.scheduledDateLocal.toLocal();
  const endDate = startDate.plus({ milliseconds: item.duration }).toLocal();
  const scheduledInterval = Interval.fromDateTimes(startDate, endDate);
```

"Invalid Interval: end before start: The end of an interval must be after its start, but you had start=2024-11-20T11:00:00.000-07:00 and end=2024-11-20T11:30:00.000-05:00"

**Comment by diesieben07** [comment]
@esetnik Can you please provide the inputs used to make the code crash? In particular, what are the exact values for `item.scheduledDateLocal` and `item.duration`? What is the system's time zone?

**Comment by esetnik** [comment]
@diesieben07 I am seeing very strange results when adding debug information:

Using the following code:

```tsx
  const startDate = item.scheduledDateLocal.toLocal();
  const endDate = startDate.plus({ milliseconds: item.duration }).toLocal();
  const systemZone = DateTime.local().zone;

  try {
    const scheduledInterval = Interval.fromDateTimes(startDate, endDate);
    return scheduledInterval;
  } catch (e) {
    console.error('scheduled interval failed', {
      scheduledDate: item.scheduledDateLocal,
      duration: item.duration,
      startDate,
      endDate,
      systemZone,
    });

    throw e;
  }
```

Produced the console log:
```
scheduled interval failed 

{
duration: 1800000,
endDate: 2024-12-09T09:30:00.000-07:00,
scheduledDate: 2024-12-09T09:00:00.000-08:00,
startDate: [Circular ~],
systemZone: {}
}
```

I have no idea why the `startDate` would be a circular reference as it's just calling `.toLocal()` on an instance of `DateTime`

**Comment by diesieben07** [comment]
@esetnik 
1. The reason it shows "Circular" is most likely because `toLocal` returns _the same object_ if the DateTime is already in the local time zone. So if `item.scheduledDateLocal` is already in the local zone (like its name suggests) then `item.scheduledDateLocal` and `startDate` are the same object. I don't know what runtime you are using, but maybe that is enough to trigger this message.
2. I am again _begging_ you to _please_ provide the input I have requested:
    - What is the exact value for `item.scheduledDateLocal`? The ISO string is not enough, as it does not include the time zone.
    - What is the system/local time zone?

   Without this information nobody can reproduce your issue and find the solution.

**Comment by esetnik** [comment]
> I am again begging you to please provide the input I have requested:

As I cannot reproduce this on my own system I have to wait for it to happen in production (only 1-2 instances per month across thousands of users). So it's hard for me to get the information you need.

> What is the exact value for item.scheduledDateLocal? The ISO string is not enough, as it does not include the time zone.

What mechanism should I use to log the `item.scheduledDateLocal` in a way that will give you all the information you need?

> What is the system/local time zone?

Again, I tried to log that and it only shows `{}` for the `systemZone`. I'm not sure how you want me to get that information.

**Comment by diesieben07** [comment]
> As I cannot reproduce this on my own system I have to wait for it to happen in production (only 1-2 instances per month across thousands of users). So it's hard for me to get the information you need.

I'm sorry, the last messages you wrote made it seem like you had reproduced it locally.

> What mechanism should I use to log the item.scheduledDateLocal in a way that will give you all the information you need?

`dt.toISO({extendedZone: true})` will include the time zone, like so: `2024-12-03T15:45:21.055-05:00[America/New_York]`.

> Again, I tried to log that and it only shows {} for the systemZone. I'm not sure how you want me to get that information.

`SystemZone.instance.name` will show the system zone regardless of whether `Settings.defaultZone` has been set.

**Comment by esetnik** [comment]
I received another instance of the issue. The test code is:

```tsx
 const startDate = item.scheduledDateLocal.toLocal();
  const endDate = startDate.plus({ milliseconds: item.duration }).toLocal();
  const systemZone = SystemZone.instance.name;

  try {
    const scheduledInterval = Interval.fromDateTimes(startDate, endDate);
    return scheduledInterval;
  } catch (e) {
    console.error('scheduled interval failed', {
      scheduledDate: item.scheduledDateLocal.toISO({ extendedZone: true }),
      duration: item.duration,
      startDate: startDate.toISO({ extendedZone: true }),
      endDate: endDate.toISO({ extendedZone: true }),
      systemZone,
    });

    throw e;
  }
```

The console log was:

```
scheduled interval failed,
{
duration: 3600000,
endDate: 2024-12-16T13:00:00.000-06:00[America/Chicago],
scheduledDate: 2024-12-16T12:00:00.000-08:00[America/Chicago],
startDate: 2024-12-16T12:00:00.000-08:00[America/Chicago],
systemZone: America/Chicago
}
```

**Comment by esetnik** [comment]
Ah glad I'm not going crazy 😝. In case it is helpful browser statistics are indicating:

- Windows >= 10 - 100% of errors
- Edge 1.31.0.0 - 61% | Chrome 129.0.0 - 39%

**Comment by esetnik** [comment]
And another user in [America/Denver]

```
{
duration: 1800000,
endDate: 2024-12-19T09:30:00.000-07:00[America/Denver],
scheduledDate: 2024-12-19T09:00:00.000-08:00[America/Denver],
startDate: 2024-12-19T09:00:00.000-08:00[America/Denver],
systemZone: America/Denver
}
```

**Comment by esetnik** [comment]
```
{
duration: 1800000,
endDate: 2025-01-22T11:30:00.000-05:00[America/New_York],
scheduledDate: 2025-01-22T11:00:00.000-06:00[America/New_York],
startDate: 2025-01-22T11:00:00.000-06:00[America/New_York],
systemZone: America/New_York
}
```

**Comment by diesieben07** [comment]
Sorry for the silence here. I _have_ found a pretty serious bug, but it alone is not enough to produce the behavior seen here, as far as I can tell.
Can you please provide the exact source for `item.scheduledDateLocal`? How is it created? Please go all the way back to a factory method like `DateTime.fromISO` and provide the inputs.

**Comment by esetnik** [comment]
`item.scheduledDateLocal` is created via:

```
// item.scheduledAt is a number in milliseconds coming from our API
const scheduledDateLocal = DateTime.fromMillis(item.scheduledAt);
```

**Comment by esetnik** [comment]
Here's an example showing the actual `scheduledAt` value that was supplied, e.g.

```
const scheduledAt = 1742493600000;
const scheduledDateLocal = DateTime.fromMillis(item.scheduledAt);
const startDate = scheduledDateLocal.toLocal();
  const endDate = startDate.plus({ milliseconds: item.duration }).toLocal();
  const systemZone = SystemZone.instance.name;

  try {
    const scheduledInterval = Interval.fromDateTimes(startDate, endDate);
    return scheduledInterval;
  } catch (e) {
    console.error('scheduled interval failed', {
      scheduledDate: item.scheduledDateLocal.toISO({ extendedZone: true }),
      duration: item.duration,
      startDate: startDate.toISO({ extendedZone: true }),
      endDate: endDate.toISO({ extendedZone: true }),
      systemZone,
    });

    throw e;
  }
```

```
{
duration: 1800000,
endDate: 2025-03-20T10:30:00.000-07:00[America/Los_Angeles],
scheduledAt: 1742493600000,
scheduledDate: 2025-03-20T10:00:00.000-08:00[America/Los_Angeles],
startDate: 2025-03-20T10:00:00.000-08:00[America/Los_Angeles],
systemZone: America/Los_Angeles
}
```

**Comment by diesieben07** [comment]
Hi @esetnik 
I just released 3.6.0 with the bugfix mentioned above. If possible, please let me know if this resolves your issue!

**Comment by esetnik** [comment]
Are you referring to the fix in https://github.com/moment/luxon/commit/20139a300ce54b280be672ada53cc9f47c8c2029?

**Comment by diesieben07** [comment]
Yes exactly.

**Comment by esetnik** [comment]
Hi @diesieben07 I just observed the same issue in v3.7.1

```
{

arguments: [
scheduled interval failed,

{
duration: 1800000,
endDate: 2025-08-14T12:30:00.000-04:00[America/New_York],
scheduledAt: 1755190800000,
scheduledDate: 2025-08-14T12:00:00.000-05:00[America/New_York],
startDate: 2025-08-14T12:00:00.000-05:00[America/New_York],
systemZone: America/New_York
}
],
logger: console
}
```

**Comment by diesieben07** [comment]
@esetnik I am sorry for the late response. I cannot explain this error except by a misbehaving `Date.prototype.getTimezoneOffset` implementation.
In your latest log you can see that `scheduledDate` is listed as `2025-10-30T12:00:00.000-06:00[America/Chicago]`. However on that date (`2025-10-30`) that time zone (`America/Chicago`) should have an offset of `-05:00`, _not_ `-06:00`.
From the [code you provided earlier](https://github.com/moment/luxon/issues/909#issuecomment-2727816288), `scheduledDate` is really just `DateTime.fromMillis(scheduledAt).toLocal()`. `toLocal` is no-op here, because `fromMillis` uses the default zone by default. There is no caching or other weirdness done by Luxon here either, the offset here comes directly out of `Date.prototype.getTimezoneOffset` for the provided epoch millis.

Unfortunately I have no other explanation and I should have come to this conclusion earlier.
To confirm this explanation I'm going to have to ask you to add another statement to the logging, which is `new Date(scheduledAt).getTimezoneOffset()`. If this also returns the wrong offset, then this is not Luxon's fault, but whatever runtime you are running on (whether the user has a buggy browser or something in your environment is patching `getTimezoneOffset` with a broken implementation).
If that returns a different offset then what Luxon reports, then I have no other explanation but pure witchcraft and a JavaScript engine which is not running the code as provided.

Also please confirm you are _not_ setting `Settings.defaultZone` - if you are, anything above is moot. However, I cannot explain you seeing this with different time zones logged unless you have left `Settings.defaultZone` alone.

**Comment by esetnik** [comment]
@diesieben07 I just got another instance of this issue. `Date.prototype.getTimezoneOffset` was `300`.

```
scheduled interval failed,

{
duration: 3600000,
endDate: 2025-12-30T13:00:00.000-05:00[America/New_York],
offset: 300,
scheduledAt: 1767124800000,
scheduledDate: 2025-12-30T12:00:00.000-08:00[America/New_York],
startDate: 2025-12-30T12:00:00.000-08:00[America/New_York],
systemZone: America/New_York
}
```
